### ISSUE TITLE
sqlmap does not detect end of output during boolean based blind retrieval

### ISSUE THREAD

**Issue by austinjdean** [sqlmapproject/sqlmap#3473]
## What's the problem (or question)?
The issue is that while sqlmap is retrieving information (e.g. length of the current user's username), during a boolean-based blind injection, it does not realize that the output is over, and continues printing spaces to stdout (retrieving character by character)

## Do you have an idea for a solution?
I have not looked at sqlmap source code very extensively, but there is most likely a break condition in the loop for extracting characters that's not getting met that should be met.

## How can we reproduce the issue?
Command: sqlmap -r request.txt --force-ssl --proxy http://127.0.0.1:8080 --delay=5 --technique B --is-dba -f --current-user --threads 5 --drop-set-cookie

The contents of request.txt are confidential.

## What are the running context details?
* Installation method (e.g. `pip`, `apt-get`, `git clone` or `zip`/`tar.gz`): Not sure. Came pre-installed.
* Client OS (e.g. `Microsoft Windows 10`): Kali Linux
* Program version (`python sqlmap.py --version` or `sqlmap --version` depending on installation): 1.3#stable
* Target DBMS (e.g. `Microsoft SQL Server`): IBM DB2
* Detected WAF/IPS protection (e.g. `ModSecurity` or `unknown`): Did not detect a WAF
* SQLi techniques found by sqlmap (e.g. `error-based` and `boolean-based blind`): boolean-based blind
* Results of manual target assessment (e.g. found that the payload `query=test' AND 4113 IN ((SELECT 'foobar'))-- qKLV` works): "123 or 1=1" returns lots of results, "123 or 1=2" returns no results.
* Relevant console output (if any): I highlighted the terminal which revealed that the spaces are green, i.e. sqlmap thinks they are valid pieces of the information.
![image](https://user-images.githubusercontent.com/7329605/52381491-b2a56200-2a3f-11e9-89b2-9b545bfadb72.png)
* Exception traceback (if any): n/a

**Comment by stamparm** [comment]
This is a known behavior on couple of DBMSes (Firebird and DB2 to be exact). Instead of getting an empty string and/or error when retrieving the character at position N+1, DBMS returns whitespace (related partially with automatic casting done by sqlmap for arbitrary data types to string-alike data)

Now, there should be a failsafe which should stop this kind of behavior after a certain number of whitespaces (https://github.com/sqlmapproject/sqlmap/blob/master/lib/core/settings.py#L188 and https://github.com/sqlmapproject/sqlmap/blob/master/lib/techniques/blind/inference.py#L633-L636)

`...it does not realize that the output is over, and continues printing spaces to stdout (retrieving character by character)` <- actually, this is wrong conclusion. sqlmap, as said in previous statement, has a failsafe mechanism for this kind of behavior. Also, in such case, it will just strip extra whitespaces from such occurrences in final results (in your case you just highlighted the console output, while internally sqlmap does the stripping)

TL;DR: Know behavior. Not a bug

**Comment by austinjdean** [comment]
Thanks for pointing out where it occurs in the code. What's the reason for the other two conditions?

`[...] and partialValue[-INFERENCE_BLANK_BREAK:].isspace() and partialValue.strip(' ')[-1:] != '\n':`

It would have stopped after 10 whitespaces were the first condition the only one, but those other two caused a snag in my case.

Despite the whitespace being stripped from the final result, this behavior still results in an arduous process of watching whitespace roll in during retrieval, seemingly endlessly. Couple that with a backend that's overloaded by more than a few requests at a time, and you've got yourself a whole lot of wasted time.

In the meantime I'll just change this locally to break after a couple whitespaces.

**Comment by stamparm** [comment]
@austinjdean please pull the latest revision and retry it again. It should be more "comfy" to run it, especially in such case you've reported

**Comment by austinjdean** [comment]
Works like a charm. Great work @stamparm
